### PR TITLE
Hcap 496 local keycloak tests

### PR DESCRIPTION
**Pull request by Zenen-freshworks**
Alright, I am _finally_ marking this feature as ready to merge. It's been a long journey to get here but the end result is relatively simple in functionality. After pulling this branch, we need to add the following entries to the .env file:
```
# For Local keycloak
KEYCLOAK_LOCAL_USERNAME=test-admin
KEYCLOAK_LOCAL_PASSWORD=password
KEYCLOAK_LOCAL_AUTH_URL=http://keycloak.local.freshworks.club:8080/auth
KEYCLOAK_LOCAL_SECRET=bf49b5d9-dd6b-46a1-b011-9ac75337cadb
```

With the .env file configured, run `make local-kc-run` and `make local-kc-run` to get the deployment properly set up. Once it's running, we can use `make kc-import-users` to import the users defined in .docker/keycloak/users.json

After the server is set up, we can open the testing console with `npx cypress open`. The example implementation of how testing works can be found in cypress/integrations/login.spec.js, and I'm looking forward to having these suites cover the breadth of our project.

An important next step will be integrating this process into the deployment workflow, but that feels like a relatively small task.

### CODE REVIEW

File: cypress/support/commands.js [hunk 1]
```diff
@@ -30,17 +46,90 @@ Cypress.Commands.add('kcGetToken', function (user) {
...
+    let client_id = Cypress.env('KEYCLOAK_API_CLIENTID');
+    let client_secret = Cypress.env('KEYCLOAK_LOCAL_SECRET');
+
+    const base64URLEncode = (str) => {
+      return str.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');
+    };
+
+    const code_challenge = base64URLEncode(crypto.randomBytes(32));
+
+    cy.request({
```
Comment (fwpushan) [inline cypress/support/commands.js:70]: As we discussed, please try direct token url
`{{keycloak_url}}/realms/{{realm}}/protocol/openid-connect/token `
HEADERS:
```json
headers [{"key":"Content-Type","name":"Content-Type","value":"application/x-www-form-urlencoded","type":"text"}]
```
body (query string / ) ->
`client_id `
`grant_type ` :  'password'
`username `
`password`
`client_secret `

File: server/keycloak.js [hunk 2]
```diff
@@ -22,13 +22,20 @@ const defaults = {
...
   constructor() {
+    const isLocal = process.env.KEYCLOAK_AUTH_URL.includes('local');
     this.realm = process.env.KEYCLOAK_REALM;
     this.authUrl = process.env.KEYCLOAK_AUTH_URL;
     this.clientNameFrontend = process.env.KEYCLOAK_FE_CLIENTID;
     this.clientNameBackend = process.env.KEYCLOAK_API_CLIENTID;
-    this.clientSecretBackend = process.env.KEYCLOAK_API_SECRET;
-    this.serviceAccountUsername = process.env.KEYCLOAK_SA_USERNAME;
-    this.serviceAccountPassword = process.env.KEYCLOAK_SA_PASSWORD;
+    this.clientSecretBackend = isLocal
```
Comment (fwpushan) [inline server/keycloak.js:30]: what is the exact requirement to have separate set of env values here?

**Review by fwpushan — COMMENTED** [review]
Great work. Just need few clarification and one improvement suggestion.

File: server/keycloak.js (hunk 2 above)
Comment (Zenen-freshworks) [reply server/keycloak.js:30]: a lot of the necessary environment variables differ between the local and remote instances. This pattern seemed like the easiest way to minimize the logic of adapting between those cases.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -21,23 +25,23 @@ jobs:
...
           wait-on: 'http://localhost:4000'
 
-      - name: Run
-        run: make local-kc-run
       - name: Sleep for 30 seconds
-        uses: jakejarvis/wait-action@master
-        with:
-          time: '30s'
+        run: sleep 30s
+        shell: bash
```
Comment (dbayly-freshworks) [inline .github/workflows/test.yml:44]: Whats the purpose of this sleep here?

File: client/package.json [hunk 4]
```diff
@@ -24,11 +24,11 @@
...
     "moment": "2.27.0",
-    "react": "16.13.1",
+    "react": "^16.13.1",
     "react-app-polyfill": "1.0.6",
-    "react-dom": "16.13.1",
+    "react-dom": "^16.13.1",
     "react-router-dom": "5.2.0",
-    "react-window": "1.8.6",
+    "react-window": "^1.8.6",
     "store": "2.0.12",
```
Comment (dbayly-freshworks) [inline client/package.json:32]: We should lock down these packages

File: server/tests/geocodes.test.js.backup [hunk 5]
```diff
@@ -0,0 +1,76 @@
+/* eslint-disable no-restricted-syntax, no-await-in-loop */
+const { v4 } = require('uuid');
+const { startDB, closeDB } = require('./util/db');
```
Comment (dbayly-freshworks) [inline server/tests/geocodes.test.js.backup:3]: If we aren't using this file we should delete it

File: cypress/support/commands.js [hunk 6]
```diff
@@ -1,46 +1,115 @@
...
+      return value;
+    }
+  }
+};
+
+const authBaseUrl = Cypress.env('KEYCLOAK_LOCAL_AUTH_URL') || Cypress.env('KEYCLOAK_AUTH_URL');
+
+Cypress.Commands.add('kcLogout', function () {
+  Cypress.log({ name: 'Logout' });
+  // let authBaseUrl = Cypress.env('KEYCLOAK_AUTH_URL');
```
Comment (Zenen-freshworks) [inline cypress/support/commands.js:24]: cleanup

File: cypress/integration/participantView.spec.js [hunk 7]
```diff
@@ -13,31 +15,33 @@ describe('Participant View', () => {
...
-    });
+    // cy.get('body').then(($body) => {
+    //   if (Cypress.$('body:contains("a day ago")').length !== 0) {
+    //     cy.contains('a day ago').should('exist');
+    //     cy.contains('14 days ago').should('exist');
+    //   } else {
+    //     cy.contains('2 days ago').should('exist');
+    //     cy.contains('15 days ago').should('exist');
+    //   }
+    // });
```
Comment (dbayly-freshworks) [inline cypress/integration/participantView.spec.js:26]: Lots of commented out code here. We should either remove it or re-add it if we need it.

File: cypress/integration/participantView.spec.js [hunk 8]
```diff
@@ -62,21 +67,23 @@ describe('Participant View', () => {
...
+  //   }).as('patchAnswer');
+  //   cy.kcLogin('test-moh');
+  //   cy.visit('/participant-view');
+  //   cy.contains('Edit').click();
+  //   cy.get('div.MuiDialog-scrollPaper').should('exist');
+  //   cy.get('input[name=firstName').should('have.value', 'Graham').clear().type('Animal');
+  //   cy.contains('Save').focus().click();
+  //   cy.wait('@patchAnswer');
+  //   cy.get('div.MuiDialog-scrollPaper').should('not.exist');
+  // });
```
Comment (dbayly-freshworks) [inline cypress/integration/participantView.spec.js:88]: More commented out code here.

File: package.json [hunk 9]
```diff
@@ -16,6 +16,7 @@
   "devDependencies": {
     "cypress": "6.1.0",
     "cypress-dotenv": "1.2.2",
+    "cypress-keycloak": "^1.7.0",
```
Comment (dbayly-freshworks) [inline package.json:19]: We should lock down this package

File: cypress/support/commands.js [hunk 10]
```diff
@@ -1,46 +1,115 @@
...
     url: `${authBaseUrl}/realms/${realm}/protocol/openid-connect/logout`,
   });
 });
 
+// Trying to phase this one out of existence
 Cypress.Commands.add('kcNavAs', function (user, visitUrl) {
   visitUrl = visitUrl || '';
   Cypress.log({ name: 'Fake Login' });
-  let authBaseUrl = Cypress.env('KEYCLOAK_AUTH_URL');
+  // let authBaseUrl = Cypress.env('KEYCLOAK_AUTH_URL');
```
Comment (dbayly-freshworks) [inline cypress/support/commands.js:112]: You can remove this line.

File: cypress/support/commands.js [hunk 11]
```diff
@@ -1,46 +1,115 @@
...
+          followRedirect: false,
+        }).its('body');
+      });
   });
 });
 
 Cypress.Commands.add('kcLogout', function () {
   Cypress.log({ name: 'Logout' });
-  let authBaseUrl = Cypress.env('KEYCLOAK_AUTH_URL');
+  // let authBaseUrl = Cypress.env('KEYCLOAK_AUTH_URL');
```
Comment (dbayly-freshworks) [inline cypress/support/commands.js:101]: You can remove this line.

File: cypress/support/commands.js [hunk 12]
```diff
@@ -1,46 +1,115 @@
...
+  });
+});
+
+Cypress.Commands.add('kcLogin', (user) => {
   Cypress.log({ name: 'Login' });
-  cy.fixture('users/' + user).then(function (userData) {
-    userData.password ? null : (userData.password = Cypress.env('KEYCLOAK_SA_PASSWORD'));
-    let authBaseUrl = Cypress.env('KEYCLOAK_AUTH_URL');
+  cy.fixture('users/' + user).then((userData) => {
+    // let authBaseUrl = Cypress.env('KEYCLOAK_AUTH_URL');
```
Comment (dbayly-freshworks) [inline cypress/support/commands.js:34]: You can remove this line.

File: cypress/integration/siteView.spec.js [hunk 13]
```diff
@@ -1,45 +1,53 @@
...
-    cy.contains('th', 'Site ID').click();
-    cy.contains('th', 'Site Name').click();
-    cy.contains('th', 'Operator Name').click();
-    cy.contains('th', 'Health Authority').click();
-    cy.contains('th', 'Postal Code').click();
+    // cy.contains('th', 'Site ID').click();
+    // cy.contains('th', 'Site Name').click();
+    // cy.contains('th', 'Operator Name').click();
+    // cy.contains('th', 'Health Authority').click();
+    // cy.contains('th', 'Postal Code').click();
```
Comment (dbayly-freshworks) [inline cypress/integration/siteView.spec.js:51]: More commented out code

**Comment by Zenen-freshworks** [comment]
There's a lot of commented out code that is there as a result of HCAP-732 being descoped from this pull request. These are valid tests that can't be implemented without seeding the participants and sites databases, so I'm uncertain about what to do with them. Open up a WIP for that HCAP-732 and keep them there, perhaps?

**Review by fwpushan — APPROVED** [review]
Great job

**Review by dbayly-freshworks — APPROVED** [review]

**Pull request merged**